## Code Review: PR #7209

### Summary

`ci.yml`'s new `renovate-config` job, the `changes` job's new `renovate` filter output, and the version pin were all independently re-verified with real command output rather than trusting the PR description's claims — including a real Linux+Node 24 container run and a pull of the actual GitHub Actions job log from this PR's own CI run. Everything the PR claims checks out. One gap exists, but it's in branch-protection configuration, not this diff — filed separately as #7216.

### What I verified

**1. Gate firing logic (push vs. PR)** — traced `changes.if: github.event_name == 'pull_request'` against `renovate-config.if: needs.changes.outputs.renovate == 'true'`. On a `push` to `main`, `changes` is skipped, and per GitHub's own Contexts reference, a skipped job's `needs.<job>.outputs.<x>` is *"An empty string if the job was skipped"* — so `'' == 'true'` is `false` and `renovate-config` cleanly **skips** (not errors). Confirmed this is safe today: `gh api repos/blamechris/chroxy/branches/main/protection` shows `enforce_admins: true` and `required_pull_request_reviews` configured, so direct pushes to `main` are already blocked — every change lands via a gated PR first.

   **However**: `gh api repos/blamechris/chroxy/branches/main/protection/required_status_checks/contexts` does **not** include `"Renovate Config"` (11 other contexts are listed, this job isn't one of them). That means even when the job runs on a PR and correctly fails, it does not block the GitHub merge button — only visibility, not enforcement, for anyone merging via the standard UI. Filed as **#7216** with the exact fix (confirmed safe: GitHub treats a *skipped* required check as passing, so adding it won't break PRs that don't touch `renovate.json`/`ci.yml`).

**2. Is the pin real?**
   ```
   $ npx --yes --package "renovate@44.30.3" -- renovate --version
   44.30.3
   $ npx --yes --package "renovate@44.30.3" -- renovate-config-validator --help
   ...
     --strict       Fail command if any configuration warnings, errors, or a migration is needed
   ...
   ```
   Pin resolves to exactly 44.30.3 (not a stale cache), and `--strict` is a real, documented flag on this exact version.

**3. Node 24 claim** — `npm view renovate@44.30.3 engines.node` → `^24.11.0`, confirmed. Pulled the **real job log** from this PR's own CI run (job 95115637071, self-hosted `chroxy-linux-winbox-01` runner):
   ```
   Acquiring 24.19.0 - x64 from https://github.com/actions/node-versions/releases/download/24.19.0-30872449280/...
   node: v24.19.0
   ```
   `actions/setup-node` with `node-version: 24` resolved 24.19.0 in production, which satisfies `^24.11.0`. Not just plausible — directly confirmed from the actual run.

   Side note for the record, not a finding: reproducing locally on macOS + Node 26 surfaced a `re2` native-addon `ERR_DLOPEN_FAILED` (ABI mismatch), which gracefully falls back to plain `RegExp` and still exits 0. Re-ran in a real `node:24-bookworm` Linux container to match the CI environment (self-hosted Linux runner + Node 24) — clean, no re2 warning, exit 0 — confirming the local warning was an artifact of my test environment (wrong OS + wrong Node major), not a real CI risk. The actual CI job log confirms the same: no re2 warning.

**4. Mutation testing** (against a copy semantics-checked before/after; real repo `renovate.json` untouched):
   | input | exit | strict required? |
   |---|---|---|
   | current `renovate.json` | **0** | — |
   | deprecated `fileMatch` key reintroduced | **1** | yes — same mutation exits **0** *without* `--strict` |
   | bogus key (`totallyBogusKey`) | **1** | fails either way |

   ```
   ERROR: Found errors in configuration
          "errors": [{"topic":"Configuration Error","message":"Invalid configuration option: totallyBogusKey"}]
   ```
   Confirmed `--strict` is load-bearing, not decorative — without it, the exact #7191 regression class (deprecated key silently accepted) passes with exit 0. `renovate.json` restored byte-identical after every mutation (`diff` empty, md5 matched, `git status --porcelain` clean throughout).

**5. Cost/latency** — real CI job (checkout + setup-node + validate): **53s** total, validate step alone **~35s**. Cold Docker repro: **23s**. Both comfortably inside the `timeout-minutes: 6` (360s) budget (~7–15x headroom); no flake risk apparent.

**6. actionlint** — ran through real `.github/workflows/` paths on both sides (not `/tmp`):
   ```
   $ actionlint .github/workflows/ci.yml   # this branch
   ci.yml:823 — label "chroxy-win" is unknown [runner-label]
   ci.yml:1060 — shellcheck SC2034: i appears unused
   ```
   Same two findings on `origin/main` (at lines 815 / 1013 — the 8-line and 47-line offsets exactly match where each of this PR's two diff hunks land relative to those lines). Both are pre-existing and unrelated to this PR's diff; no new findings introduced. Also checked for a second `renovate.json`/`renovate.json5` elsewhere in the tree that the filter might be missing — there is only the one file, and it's the one being watched.

### Issues Found

**Critical:** none in this diff.

**Suggestions:**
- (Filed as **#7216**, not blocking this PR) Add `"Renovate Config"` to the branch protection `required_status_checks.contexts` list so a failing validation actually blocks merge via the GitHub UI, not just via this repo's own autonomous "all CI green" convention.
- Since this repo already has a `customManagers` regex tracker for the sidecar Dockerfile's pinned `CLAUDE_CODE_VERSION`, consider adding a similar one for `RENOVATE_VERSION` in `ci.yml` so Renovate can propose its own version bumps as a PR (which this new gate would then validate) instead of the pin only being bumped by hand.

**Nitpicks:**
- Unrelated to this PR's diff, noticed in passing: the `changes` job's `platform` filter output (used by the pre-existing filter block above the new `renovate` one) no longer appears to gate anything — `server-tests-windows` dropped its path-based `if:` when it moved to the self-hosted runner (per that job's own comment), so `needs.changes.outputs.platform` has no remaining consumer. Not something to fix here; flagging only because this PR's diff sits right next to it.

### Verdict

**Approve.** The workflow logic, the version pin, the Node 24 requirement, and the `--strict` flag all check out under direct adversarial testing (real mutation tests, a real Linux+Node24 container, and the actual CI job log from this PR). The only gap found — the check not being in required branch-protection contexts — is a repository-settings change outside this diff's scope, tracked as #7216.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -900,6 +900,13 @@ jobs:
           # something a validation gate should be built on. Bump alongside
           # RENOVATE_VERSION below.
           node-version: 24
+          # Reuse ~/.npm across runs so the validator's dependency tree is not
+          # re-downloaded cold every time. Safe alongside the pin below: npx
+          # resolves an explicit `renovate@<version>` spec regardless of what
+          # is cached — the staleness that motivated pinning came from an
+          # UNPINNED spec silently reusing whatever the cache held.
+          cache: npm
+          cache-dependency-path: package-lock.json
 
       - name: Validate renovate.json
         env:

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -178,6 +178,7 @@ jobs:
       pull-requests: read
     outputs:
       platform: ${{ steps.filter.outputs.platform }}
+      renovate: ${{ steps.filter.outputs.renovate }}
     steps:
       - uses: actions/checkout@34e114876b0b11c390a56381ad16ebd13914f8d5 # v4
 
@@ -198,6 +199,13 @@ jobs:
               - 'packages/server/tests/platform*.test.js'
               # Workflow itself — so edits to this gating logic always
               # exercise the gated job on the same PR.
+              - '.github/workflows/ci.yml'
+            renovate:
+              # #7196 — renovate.json is exercised by nothing else in CI, so an
+              # edit can reintroduce invalid or deprecated syntax with nothing
+              # catching it before merge. That is exactly how it accumulated the
+              # deprecated regexManagers/fileMatch spelling #7191 had to fix.
+              - 'renovate.json'
               - '.github/workflows/ci.yml'
 
   server-tests-windows:
@@ -861,6 +869,45 @@ jobs:
 
       - name: Run desktop tests
         run: cargo test --locked
+
+  renovate-config:
+    name: Renovate Config
+    # #7196 — validate renovate.json against a PINNED renovate.
+    #
+    # The version pin is the point. `npx --package renovate` with no version
+    # reuses whatever is already in the local ~/.npm/_npx cache rather than
+    # necessarily re-resolving, so the same command can validate against a
+    # months-old renovate and reject a correct config — which is exactly what
+    # happened while reviewing #7191 (a cached 37.x rejected the current
+    # `managerFilePatterns` spelling). A gate that fails on correct input, or
+    # passes on stale rules, is worse than no gate.
+    #
+    # Bumping this pin is deliberate work: renovate's schema moves, and the
+    # bump should be the change that surfaces any migration, not a silent
+    # drift underneath an unpinned command.
+    needs: [runner-target, changes]
+    if: needs.changes.outputs.renovate == 'true'
+    runs-on: ${{ fromJSON(needs.runner-target.outputs.runner) }}  # SELF_HOSTED_OR_HOSTED (see header)
+    timeout-minutes: 6
+    steps:
+      - uses: actions/checkout@34e114876b0b11c390a56381ad16ebd13914f8d5 # v4
+
+      - uses: actions/setup-node@49933ea5288caeca8642d1e84afbd3f7d6820020 # v4
+        with:
+          # 24, not the repo's usual 22: renovate 44.x declares
+          # `engines.node: ^24.11.0`, and running it under 22 only produces an
+          # EBADENGINE warning today — an unsupported-engine combination is not
+          # something a validation gate should be built on. Bump alongside
+          # RENOVATE_VERSION below.
+          node-version: 24
+
+      - name: Validate renovate.json
+        env:
+          # Keep in sync deliberately; see the note above before changing.
+          RENOVATE_VERSION: 44.30.3
+        run: |
+          npx --yes --package "renovate@${RENOVATE_VERSION}" -- \
+            renovate-config-validator --strict renovate.json
 
   scripts-tests:
     name: Scripts Tests

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -906,7 +906,11 @@ jobs:
           # is cached — the staleness that motivated pinning came from an
           # UNPINNED spec silently reusing whatever the cache held.
           cache: npm
-          cache-dependency-path: package-lock.json
+          # Glob, not the bare root lockfile: this monorepo has three
+          # (root, packages/server, packages/server/sidecar) and the cache key
+          # must cover all of them. packages/server/tests/ci-cache-key.test.js
+          # enforces this for every job in this file.
+          cache-dependency-path: '**/package-lock.json'
 
       - name: Validate renovate.json
         env:

